### PR TITLE
plugins/auto-session: init + tests

### LINKED ISSUE #426: Plugin request: auto-session
**Plugin**:  auto-session
**Plugin repository**: https://github.com/rmagatti/auto-session
**Is the plugin on nixpkgs**: https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=auto-session

## Extra information
Session Lens was once its own plugin, so it has its own enable/settings/etc. Thank you!

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for the [`auto-session`](https://github.com/rmagatti/auto-session) plugin.

Fixes #426.

### CODE REVIEW

File: plugins/utils/auto-session.nix [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    helpers.extraOptionsOptions
+    // {
+      enable = mkEnableOption "auto-session";
+
+      package = helpers.mkPackageOption "auto-session" pkgs.vimPlugins.auto-session;
+
+      logLevel =
+        helpers.defaultNullOpts.mkEnum
+        ["debug" "info" "warn" "error"]
+        "error"
```
Comment (traxys) [inline plugins/utils/auto-session.nix:21]: Upstream default seems to be info ?

File: plugins/utils/auto-session.nix [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        "Sets the log level of the plugin.";
+
+      autoSession = {
+        enable = helpers.defaultNullOpts.mkBool true ''
```
Comment (traxys) [inline plugins/utils/auto-session.nix:25]: Default is nil. Do we leave `enable` when the upstream option is `enabled` ?

File: plugins/utils/auto-session.nix [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        rootDir =
+          helpers.defaultNullOpts.mkNullable
+          (with types; either str helpers.rawType)
+          "{__raw = \"vim.fn.stdpath 'data' .. '/sessions/'\";}"
+          ''
+            Root directory for session files.
+            Can be either a string or lua code (using `{__raw = 'foo';}`).
+          '';
+
+        createEnabled = helpers.mkNullOrOption types.bool ''
```
Comment (traxys) [inline plugins/utils/auto-session.nix:42]: default is true

File: plugins/utils/auto-session.nix [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        '';
+
+        useGitBranch = helpers.mkNullOrOption types.bool ''
+          Include git branch name in session name to differentiate between sessions for different
+          git branches.
+        '';
+      };
+
+      autoSave = {
+        enable = helpers.defaultNullOpts.mkNullable types.bool "null" ''
```
Comment (traxys) [inline plugins/utils/auto-session.nix:61]: do we leave `enable` when the upstream option is `enabled` ?

File: plugins/utils/auto-session.nix [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+          Whether to enable auto saving session.
+        '';
+      };
+
+      autoRestore = {
+        enable = helpers.defaultNullOpts.mkNullable types.bool "null" ''
```
Comment (traxys) [inline plugins/utils/auto-session.nix:67]: `enable` or `enabled` ?

File: plugins/utils/auto-session.nix [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+          Whether to enable auto restoring session.
+        '';
+      };
+
+      cwdChangeHandling =
+        helpers.defaultNullOpts.mkNullable
+        (
+          with types;
+            either
+            str
```
Comment (traxys) [inline plugins/utils/auto-session.nix:77]: Why can this be a str ?

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/auto-session.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/utils/auto-session.nix:21]: This line suggests `"error"` no ?
https://github.com/rmagatti/auto-session/blob/7afbb149f87be279778689596c781882014f7eef/lua/auto-session/init.lua#L52

Comment (traxys) [reply plugins/utils/auto-session.nix:21]: It seems like, the readme was telling `info` though....

File: plugins/utils/auto-session.nix (hunk 5 above)
Comment (GaetanLepage) [reply plugins/utils/auto-session.nix]: I have changed to `enabled` to stick to the upstream naming.

File: plugins/utils/auto-session.nix (hunk 2 above)
Comment (GaetanLepage) [reply plugins/utils/auto-session.nix]: Default is actually `true`.
Renamed to `enabled`.

File: plugins/utils/auto-session.nix (hunk 3 above)
Comment (GaetanLepage) [reply plugins/utils/auto-session.nix:42]: Default is actuall `nil`.

File: plugins/utils/auto-session.nix (hunk 6 above)
Comment (GaetanLepage) [reply plugins/utils/auto-session.nix]: My mistake. It can either be a table **or `false`**. I fixed it.

**Review by traxys — APPROVED** [review]

**Pull request merged**